overlord/assertstate/assertstate.go

```diff
@@ -911,6 +920,61 @@ func ApplyEnforcedValidationSets(st *state.State, valsets map[string]*asserts.Va
 		valsetsTracking = append(valsetsTracking, tr)
 		valsetsSlice = append(valsetsSlice, vs)
 	}
+	return valsetsSlice, valsetsTracking, nil
+}
+
+// ApplyLocalEnforcedValidationSets enforces the supplied validation sets. It takes a map
+// of validation set keys to validation sets, pinned sequence numbers (if any),
+// installed snaps and ignored snaps. The local in this naming indicates that it uses the
+// validation-set primary keys to lookup assertions in the current database. No fetching is
+// done contrary to the non-local version.
+func ApplyLocalEnforcedValidationSets(st *state.State, vsKeys map[string][]string, pinnedSeqs map[string]int, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool) error {
+	valsets, err := resolveValidationSetPrimaryKeys(st, vsKeys)
+	if err != nil {
+		return err
+	}
+
+	valsetsSlice, valsetsTracking, err := validationSetTrackings(valsets, pinnedSeqs)
+	if err != nil {
+		return err
+	}
+
+	valsetGroup, err := TrackedEnforcedValidationSets(st, valsetsSlice...)
+	if err != nil {
+		return err
+	}
+
+	if err := valsetGroup.Conflict(); err != nil {
```

Comment:
looking at this reminds me that you added Writer.CheckValidationSets, but do we call it from somewhere during image building?

Reply:
For some reason this went unnoticed by me, I think I closed a PR where this was done. Introduced here https://github.com/snapcore/snapd/pull/12763